### PR TITLE
Fix #43295, better handling of consul initialization

### LINKED ISSUE #43295: salt.cache.redis_cache does not actually work.
According to [this page](https://docs.saltstack.com/en/latest/ref/cache/all/salt.cache.redis_cache.html#module-salt.cache.redis_cache), I should be able use a redis server to store the minion data cache in order to share it between different masters. But... it doesn't actually work. The docs page helpfully does not actually include a complete configuration example because who needs to be thorough, but I put the following configuration together based on what little information is offered in the documentation. And, yes, python-redis is properly installed.

master configuration:
```
redis_cache.host: <redis server>
redis_cache.port: <redis port>
redis_cache.password: <redis_password>
redis_cache.db: '0'
cache: redis_cache
```

But when running `salt-run cache.grains`, the command crashes with this:

error:
```
[root@psc-saltmaster1 ~]# salt-run cache.grains
[ERROR   ] Exception raised when processing __virtual__ function for salt.loaded.int.cache.consul. Module will not be loaded: 'module' object has no attribute 'Consul'
Exception occurred in runner cache.grains: Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/client/mixins.py", line 392, in _low
    data['return'] = self.functions[fun](*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/salt/runners/cache.py", line 60, in grains
    cached_grains = pillar_util.get_minion_grains()
  File "/usr/lib/python2.7/site-packages/salt/utils/master.py", line 312, in get_minion_grains
    cached_minion_grains, cached_minion_pillars = self._get_cached_minion_data(*minion_ids)
  File "/usr/lib/python2.7/site-packages/salt/utils/master.py", line 144, in _get_cached_minion_data
    minion_ids = self.cache.ls('minions')
  File "/usr/lib/python2.7/site-packages/salt/cache/__init__.py", line 244, in ls
    return self.modules[fun](bank, **self._kwargs)
  File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1113, in __getitem__
    func = super(LazyLoader, self).__getitem__(item)
  File "/usr/lib/python2.7/site-packages/salt/utils/lazy.py", line 101, in __getitem__
    raise KeyError(key)
KeyError: 'redis_cache.ls'
```

It's funny that it's trying to open Consul, because last time I checked, `redis_cache` is not equal to `consul`, but what do I know. I did install `pyconsul` as well just in case, but that had no effect on the issue.

version report:
```
Salt Version:
           Salt: 2017.7.1
 
Dependency Versions:
           cffi: 1.6.0
       cherrypy: 5.6.0
       dateutil: 1.5
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: 0.25.1
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: 2.14
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: 0.25.1
         Python: 2.7.5 (default, Nov  6 2016, 00:28:07)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4
 
System Versions:
           dist: centos 7.3.1611 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-514.21.1.el7.x86_64
         system: Linux
        version: CentOS Linux 7.3.1611 Core
```

### PR DESCRIPTION
**Pull request by johnj**
### What does this PR do?

Handle consul.Consul throwing an AttributeError exception

### What issues does this PR fix or reference?

#43295

### CODE REVIEW

File: salt/cache/consul.py [hunk 1]
```diff
@@ -81,8 +81,11 @@ def __virtual__():
             'verify': __opts__.get('consul.verify', True),
             }
 
-    global api
-    api = consul.Consul(**consul_kwargs)
+    try:
+        global api
+        api = consul.Consul(**consul_kwargs)
+    except AttributeError:
+        return (False, "Failed to invoke consul.Consul, please make sure you have python-consul >= 0.2.0 installed")
```
Comment (rallytime) [inline salt/cache/consul.py]: @johnj Can you add this requirement to the docs near the top of this file for the lib requirement as well?

Just something like:
```
:depends: python-consul >= 0.2.0
```
That will help people trouble shoot this as well.

**Review by rallytime — COMMENTED** [review]
Thanks for adding this check, @johnj. I just had one minor addition that we should add here as well.

Once that is added, we can get this in. :)

File: salt/cache/consul.py (hunk 1 above)
Comment (johnj) [reply salt/cache/consul.py]: Thanks @rallytime, incoming update 😄

**Review by rallytime — APPROVED** [review]

**Review by cachedout — APPROVED** [review]

**Comment by gtmanfred** [comment]
This needs to be backported to 2016.11

**Pull request merged**